climate/plugins/oshosts/host_plugin.py

```diff
@@ -124,52 +126,57 @@ def create_reservation(self, values, usage_enforcement=False, usage_db_host=None
             'status': 'pending',
         }
 
+        host_reservation = db_api.host_reservation_create(host_values)
+        host_ids = self._matching_hosts(
+            values['hypervisor_properties'],
+            values['resource_properties'],
+            count_range,
+            values['start_date'],
+            values['end_date'],
+        )
+        if not host_ids:
+            pool.delete(pool_name)
+            raise manager_ex.NotEnoughHostsAvailable()
+
+
         # Check if we have enough available SUs for this reservation
         if usage_enforcement:
+            total_su_factor = sum(billrate.computehost_billrate(host_id) for host_id in host_ids)
             try:
                 balance = float(r.hget('balance', project_name))
                 encumbered = float(r.hget('encumbered', project_name))
                 start_date = values['start_date']
                 end_date = values['end_date']
                 duration = end_date - start_date
-                hours = (duration.days * 86400 + duration.seconds) / 3600.0
-                requested = hours * float(values['max'])
+                hours = duration.total_seconds() / 3600.0
+                requested = hours * total_su_factor
                 left = balance - encumbered
                 if left - requested < 0:
                     raise common_ex.NotAuthorized(
                         'Reservation for project %s would spend %f SUs, only %f left' % (project_name, requested, left))
             except redis.exceptions.ConnectionError:
                 LOG.exception("cannot connect to redis host %s", CONF.manager.usage_db_host)
 
-        db_api.host_reservation_create(host_values)
-        host_ids = self._matching_hosts(
-            values['hypervisor_properties'],
-            values['resource_properties'],
-            count_range,
-            values['start_date'],
-            values['end_date'],
-        )
-        if not host_ids:
-            pool.delete(pool_name)
-            raise manager_ex.NotEnoughHostsAvailable()
-        for host_id in host_ids:
-            db_api.host_allocation_create({'compute_host_id': host_id,
-                                          'reservation_id': reservation['id']})
-
-        if usage_enforcement:
             try:
```

Comment:
If your testing shows it's fine to move the block calling host_reservation_create higher up as you did, you can merge the two try/except blocks interacting with Redis into a single one.